### ISSUE TITLE
[ES|QL] Materialized views

### ISSUE THREAD

**Issue by tylerperk** [elastic/elasticsearch#108640]
### Description

TL;DR: We might want to build something similar to Materialized Views in ES|QL. This can have multiple benefits: 

- encapsulate complex queries to be easily reused
- pre-calculate joins and aggregations for more performance when they are actually needed
- pre-populate lists of values (usernames, IP addresses, etc) to be used for query-time enrichment
- Being able to persist an ES|QL query with a name and use it in the FROM command for follow up calls would make building some of the UIs simpler as often the same pre-filtering of data happens multiple times
- and more

A nice description from [1129](https://github.com/elastic/ml-team/issues/1129): "To keep query complexity under control and allow reuse of complex query components, SQL databases have the concept of views. A view is the output of an SQL query run against some base tables. Then the view can be queried using a separate SQL query. The database engine effectively combines the two, but the outermost query can be very simple, with complex joins hidden from that final query because they're in the view definition.

Since view definitions often contain complex joins to hide them away, this potentially creates a performance problem. A user could query a view with a seemingly very simple query, yet it takes an eternity to run because behind the scenes many tables are being joined. To make matters worse, many end users may want to query the same view, and each of them is causing the heavyweight join processing to happen for every query they run. To solve this problem databases incorporated the concept of materialized views. The idea here is that the output of the view definition is persisted in a secret table so that queries against the view don't have to redo all the work of the view definition query."

**Comment by jasonrhodes** [comment]
@dimkots I know it's been a while but is there an update on anything related to this, at all?

**Comment by uri-weisman** [comment]
@tylerperk Our team, which is responsible for the Entity Store (a centralized infrastructure for security entities), is very interested in MV as a potential replacement for our current use of transforms.
We’d like to learn more about the current status of the project.

**Comment by benwtrent** [comment]
A handful of very difficult things IMO for materialized views:

 - State management needs to be customizable. If we (ML) want to be able to use materialized views to apply to anomaly detection jobs (continuous or batch), ML needs to be able to apply state changes/actions on MV run times. Maybe this is as simple as having a hook on MV task start that ML can hook into. Just wanting to call this out.
 - MV running as a persistent task that can survive nodes leaving the cluster
 - Where to store the results of the MV

These are some early cautions (that y'all probably already thought of)

**Comment by getkub** [comment]
This is a very important feature. Any updates on this please?

**Comment by wchaparro** [comment]
We're in the design phase / POC phase for materialized views now.  We've just completed the work necessary to launch logical views into technical preview in an upcoming serverless release.
